### PR TITLE
Detector-Competition-Feat -  Dockerhub detector

### PR DESCRIPTION
**Pull request by ankushgoel27**
Opening this PR again as PR #1863 was closed due to inactivity.

Description:
This PR adds the older format of dockerhub token which is still valid and i have seen it in the wild. add the "id" to username regex.

Reference :
https://www.docker.com/blog/docker-hub-new-personal-access-tokens/

dockerhub

### CODE REVIEW

**Comment by zricethezav** [comment]
@ankushgoel27 thanks for submitting this PR. We're still considering it and will get a review on it soon. If merged, it'll be considered for the detector competition 👍🏻

**Comment by ankushgoel27** [comment]
@zricethezav i believe this is also ready for merge

**Comment by ankushgoel27** [comment]
can i get a review here please

File: pkg/detectors/dockerhub/dockerhub.go [hunk 1]
```diff
@@ -13,26 +13,29 @@ import (
...
+	defaultClient = common.SaneHttpClient()
 
 	// Can use email or username for login.
-	usernamePat = regexp.MustCompile(`(?im)(?:user|usr|-u)\S{0,40}?[:=\s]{1,3}[ '"=]?([a-zA-Z0-9]{4,40})\b`)
+	usernamePat = regexp.MustCompile(`(?im)(?:user|usr|-u|id)\S{0,40}?[:=\s]{1,3}[ '"=]?([a-zA-Z0-9]{4,40})\b`)
 	emailPat    = regexp.MustCompile(common.EmailPattern)
 
 	// Can use password or personal access token (PAT) for login, but this scanner will only check for PATs.
-	accessTokenPat = regexp.MustCompile(`\bdckr_pat_([a-zA-Z0-9_-]){27}\b`)
+	accessTokenPat = regexp.MustCompile(detectors.PrefixRegex([]string{"docker"}) + `\b([0-9Aa-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})\b`)
```
Comment (rosecodym) [inline pkg/detectors/dockerhub/dockerhub.go:32]: `[0-9Aa-f]` is a pretty unusual character group - why is `A` the only included capital letter? (If it was an accident, can you just reuse a pattern in common/patterns.go?)

File: pkg/detectors/dockerhub/dockerhub.go [hunk 2]
```diff
@@ -54,14 +57,23 @@ func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (result
 
 	for _, resUserMatch := range userMatches {
 		for _, resAccessTokenMatch := range accessTokenMatches {
+			if len(resAccessTokenMatch) != 2 {
+                        	continue
+                	}
+                	pat := strings.TrimSpace(resAccessTokenMatch[1])
```
Comment (rosecodym) [inline pkg/detectors/dockerhub/dockerhub.go:63]: Why is trimming necessary here? I don't see any way to capture whitespace in the access token pattern. Am I misreading it?

File: pkg/detectors/dockerhub/dockerhub.go [hunk 3]
```diff
@@ -54,14 +57,23 @@ func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (result
...
 
 			s1 := detectors.Result{
 				DetectorType: detectorspb.DetectorType_Dockerhub,
-				Raw:          []byte(fmt.Sprintf("%s: %s", resUserMatch, resAccessTokenMatch)),
+				Raw:          []byte(fmt.Sprintf("%s: %s", resUserMatch, pat)),
```
Comment (rosecodym) [inline pkg/detectors/dockerhub/dockerhub.go:67]: please also set `RawV2` with the new value

File: pkg/detectors/dockerhub/dockerhub.go [hunk 4]
```diff
@@ -79,8 +91,15 @@ func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (result
 					// Valid credentials can still return a 401 status code if 2FA is enabled
 					if (res.StatusCode >= 200 && res.StatusCode < 300) || (res.StatusCode == 401 && strings.Contains(string(body), "login_2fa_token")) {
 						s1.Verified = true
+					} else if res.StatusCode == 401 {
+                                        // The secret is determinately not verified (nothing to do)
+                                	} else {
+						s1.VerificationError = fmt.Errorf("unexpected HTTP response status %d", res.StatusCode)
```
Comment (rosecodym) [inline pkg/detectors/dockerhub/dockerhub.go:97]: This isn't actually possible any more - you need to use [a function](https://github.com/trufflesecurity/trufflehog/blob/main/pkg/detectors/detectors.go#L68) to set these errors now. You'll need to resolve this when you update your base branch.

File: pkg/detectors/dockerhub/dockerhub.go [hunk 5]
```diff
@@ -13,26 +13,29 @@ import (
...
 )
 
-type Scanner struct{}
+type Scanner struct{
+	client *http.Client
+}
+
 
-// Ensure the Scanner satisfies the interface at compile time.
+// Ensure the Scanner satisfies the interfaces at compile time.
```
Comment (rosecodym) [inline pkg/detectors/dockerhub/dockerhub.go:21]: If you're going to add a version 2 of this detector you should also make version 1 implement `Versioner` and report its version as 1.

File: pkg/detectors/dockerhub_v2/dockerhub_v2.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+					defer res.Body.Close()
+					body, err := io.ReadAll(res.Body)
+					if err != nil {
+						continue
+					}
+
+					// Valid credentials can still return a 401 status code if 2FA is enabled
+					if (res.StatusCode >= 200 && res.StatusCode < 300) || (res.StatusCode == 401 && strings.Contains(string(body), "login_2fa_token")) {
+						s1.Verified = true
+					}
```
Comment (rosecodym) [inline pkg/detectors/dockerhub_v2/dockerhub_v2.go:84]: If we're touching this detector we should add indeterminacy support while we're here.

File: pkg/detectors/dockerhub_v2/dockerhub_v2.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
+package dockerhub_v2
```
Comment (rosecodym) [inline pkg/detectors/dockerhub_v2/dockerhub_v2.go:1]: It looks like the other versioned detectors don't use underscores in their package names. Can you do the same here for consistency? (i.e. `dockerhubv2`)

**Review by rosecodym — COMMENTED** [review]
I left some inline comments, and you're going to have to update your base branch (sorry for the review delay :( ), but taking a step back, I have a question: Why did you make the existing detector v2 and add a "new" v1 rather than just adding the new detector as v2?

File: pkg/detectors/dockerhub_v2/dockerhub_v2.go (hunk 7 above)
Comment (ankushgoel27) [reply pkg/detectors/dockerhub_v2/dockerhub_v2.go:1]: it says to do so here - add _v2 to package name - https://github.com/trufflesecurity/trufflehog/blob/main/hack/docs/Adding_Detectors_external.md

**Comment by ankushgoel27** [comment]
i created this new PR - https://github.com/trufflesecurity/trufflehog/pull/2361. This PR was stale and so many changes has happened since then in how the detectors work.

File: pkg/detectors/dockerhub_v2/dockerhub_v2.go (hunk 7 above)
Comment (rosecodym) [reply pkg/detectors/dockerhub_v2/dockerhub_v2.go:1]: That is an excellent point. I'll make a note to get that cleaned up, thanks!

Comment (rgmz) [reply pkg/detectors/dockerhub_v2/dockerhub_v2.go:1]: Hot take: underscores make package names easier to read.

**Comment by zricethezav** [comment]
closing as https://github.com/trufflesecurity/trufflehog/pull/2361 has been merged

**Pull request closed without merging**